### ISSUE TITLE
oauthclient: Protect against exceptions when handling oauth2 responses

### ISSUE THREAD

**Issue by emanueldima** [inveniosoftware/invenio#3029]
While testing a new oauth2 server I got an exception from the oauthclient module. The server I'm testing  fails to return the authorization token and instead returns a 500 HTTP error with XML content but without setting the content-type. The flask oauthlib assumes it's json and tries to read it, thus triggering an exception. I think it doesn't make sense to change the flask library, but Invenio should protect against this case.

TL;DR: display a nice error message when this line
https://github.com/inveniosoftware/invenio/blob/master/invenio/modules/oauthclient/views/client.py#L182
throws an exception

**Comment by lnielsen** [comment]
Hi, 

Do you have a stack trace I can take a quick look? I agree oauthclient should protect against bad handlers in the line you have mentioned. I'd however like to understand where in oauthlib the problem is.

**Comment by emanueldima** [comment]
Yes, here it is:

````
Traceback (most recent call last):
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/vagrant/.virtualenvs/b2share/src/b2share/invenio/ext/legacy/__init__.py", line 128, in __call__
    response = self.app.handle_exception(e)
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/flask_restful/__init__.py", line 265, in error_router
    return original_handler(e)
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/flask/app.py", line 1403, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/vagrant/.virtualenvs/b2share/src/b2share/invenio/ext/legacy/__init__.py", line 124, in __call__
    response = self.app.full_dispatch_request()
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/flask_restful/__init__.py", line 265, in error_router
    return original_handler(e)
  File "/home/vagrant/.virtualenvs/b2share/src/b2share/invenio/base/wrappers.py", line 125, in handle_user_exception
    return super(Flask, self).handle_user_exception(e)
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/vagrant/.virtualenvs/b2share/src/b2share/invenio/modules/oauthclient/views/client.py", line 185, in authorized
    return handlers[remote_app]()
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/flask_oauthlib/client.py", line 646, in decorated
    data = self.handle_response()
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/flask_oauthlib/client.py", line 631, in handle_response
    data = self.handle_oauth2_response()
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/flask_oauthlib/client.py", line 610, in handle_oauth2_response
    data = parse_response(resp, content, content_type=self.content_type)
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/flask_oauthlib/client.py", line 120, in parse_response
    return json.loads(content)
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/flask/json.py", line 149, in loads
    return _json.loads(s, **kwargs)
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/simplejson/__init__.py", line 522, in loads
    return cls(encoding=encoding, **kw).decode(s)
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/home/vagrant/.virtualenvs/b2share/lib/python2.7/site-packages/simplejson/decoder.py", line 400, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
JSONDecodeError: Expecting value: line 1 column 1 (char 0)
````

**Comment by jirikuncar** [comment]
(a) register generic error handler for OAuth Client urls using `@blueprint.errorhandler(Exception)` (after https://github.com/inveniosoftware/invenio/blob/master/invenio/modules/oauthclient/views/client.py#L38)
(b) for `JSONDecodeError` we should probably abort with 400

**Comment by lnielsen** [comment]
Would go for (b)

**Comment by jirikuncar** [comment]
What about returning 502? 

See http://tools.ietf.org/html/rfc7231#section-6.6.3:
> **`6.6.3.  502 Bad Gateway`**
> 
>     The 502 (Bad Gateway) status code indicates that the server, while
>     acting as a gateway or proxy, received an invalid response from an
>     inbound server it accessed while attempting to fulfill the request.

OAuthClient is acting as a gateway to `authorize_url` that returns wrong response.

WDYT @emanueldima @lnielsen

**Comment by emanueldima** [comment]
I agree with 502, makes sense to me. Will the error have a html payload, with a nice error message to the user?

**Comment by jirikuncar** [comment]
Invenio doesn't have nice 502 error page by default. Maybe it's time to create one ...

**Comment by lnielsen** [comment]
I would use 400 (bad request) or alternatively 415 (unsupported media type) instead.

5xx errors are server errors, where as 4xx errors are client errors. In this case, it's a client side error since ``Content-Type`` is not set in the request, and thus the request body is assumed to be ``application/json``. The client should have set ``Content-Type`` to ``application/xml`` in which case the request body would be parsed correctly.

415 would be best suited if we require ``Content-Type`` to be set and thus don't have a default value (but this might break existing integrations). So probably a 400 with a proper error message is the best.

**Comment by lnielsen** [comment]
Actually, just talked to @jirikuncar who pointed out the error is when the server tries to exchange the one time code for an access token.

According to the traceback here:
https://github.com/lepture/flask-oauthlib/blob/v0.6.0/flask_oauthlib/client.py#L610

Actually, I think this is a configuration mistake that can be solved by setting ``content_type`` in the unit's remote app. See https://github.com/lepture/flask-oauthlib/blob/v0.6.0/flask_oauthlib/client.py#L207

Alternatively unity server should return a proper content-type.

In any case the current code would be correct in returning a 500 error.

**Comment by lnielsen** [comment]
I.e. here https://github.com/EUDAT-B2SHARE/b2share/blob/master/invenio/modules/oauthclient/contrib/unity.py#L62 I think you should add:

```python
app_key="UNITY_APP_CREDENTIALS",
content_type="application/xml"
```

**Comment by emanueldima** [comment]
The Unity server does not set the content type when returning an error with xml payload, and that's bad. But all the normal communication is json.
